dali/operators/reader/loader/webdataset_loader.cc

```diff
@@ -0,0 +1,369 @@
+}
+
+inline Index WebdatasetLoader::GetCurrentSampleIndex() const {
+  return wds_shards_prefixsums_[current_wds_shard_index_] + current_sample_index_;
+}
+
+inline size_t WebdatasetLoader::MaxCommonDtypeSize(const std::string& extension) const {
+  size_t dtype_max_size = 1;
+  for (auto& component_index : ext_map_.at(extension)) {
+    dtype_max_size =
+        std::max(dtype_max_size, TypeTable::GetTypeInfo(dtype_[component_index]).size());
+  }
+  return dtype_max_size;
+}
+
+inline void WebdatasetLoader::SetDataPointer(std::vector<Tensor<CPUBackend>>& sample,
+                                             std::vector<bool>& sample_was_set,
+                                             const std::string& extension,
+                                             const std::string& source_info,
+                                             std::shared_ptr<void> data, int64_t size) const {
+  DALIMeta meta;
+  meta.SetSourceInfo(source_info);
+  meta.SetSkipSample(false);
+
+  for (size_t component_index : ext_map_.at(extension)) {
+    if (!sample_was_set[component_index]) {
+      auto component_dtype_info = TypeTable::GetTypeInfo(dtype_[component_index]);
+      sample[component_index].SetMeta(meta);
+      sample[component_index].ShareData(data, size,
+                                        {size / static_cast<int64_t>(component_dtype_info.size())},
+                                        component_dtype_info);
+      sample_was_set[component_index] = true;
+    }
+  }
+}
+
+
+inline uint8_t* WebdatasetLoader::ShareDataPointer(std::vector<Tensor<CPUBackend>>& sample,
+                                                   std::vector<bool>& sample_was_set,
+                                                   const std::string& extension,
+                                                   const std::string& source_info, int64_t size,
+                                                   size_t dtype_max_size) const {
+  DALIMeta meta;
+  meta.SetSourceInfo(source_info);
+  meta.SetSkipSample(false);
+
+  uint8_t* shared_tensor_data = nullptr;
+  for (size_t component_index : ext_map_.at(extension)) {
+    if (!sample_was_set[component_index]) {
+      sample[component_index].SetMeta(meta);
+      auto component_dtype_info = TypeTable::GetTypeInfo(dtype_[component_index]);
+      if (shared_tensor_data == nullptr) {
+        sample[component_index].reserve(align_up(size, dtype_max_size));
+        sample[component_index].Resize(size / component_dtype_info.size(), component_dtype_info);
+        shared_tensor_data = reinterpret_cast<uint8_t*>(sample[component_index].raw_mutable_data());
+        std::memset(shared_tensor_data + size, 0, align_up(size, dtype_max_size) - size);
+      } else {
+        sample[component_index].ShareData(
+            shared_tensor_data, size, {size / static_cast<int64_t>(component_dtype_info.size())},
+            component_dtype_info);
+      }
+      sample_was_set[component_index] = true;
```

Comment:
yes, it should

Reply:
I would add it for the readability.

Reply:
add what? break would mean a behavior that we would not want

Reply:
Mhh, so why we share the same memory with different `sample[component_index]`? I would expect to get a different piece every time.

Reply:
no, because what ext_map_ does is it maps a specific extension to which indicies of samples it should go to, so those components are actually meant to share the same data

Reply:
Understood.

Reply:
I'm having second thought about that. How often we want components to share the underlying memory?
Would be something like:
```
["a.a;a.b;a.a;a.b", "b.a;b.b"],
```
?

Reply:
well that's just one output so no it wouldn't